### PR TITLE
Item upgrades v2 - turnin handling

### PR DESCRIPTION
**Pull request by SabreValkyrn**

### CODE REVIEW

File: ItemUpgrades.lua [hunk 1]
```diff
@@ -775,9 +775,10 @@ function addon.itemUpgrades:GetItemData(itemLink, tooltip)
 
     local itemData = {
         itemLink = itemLink,
-        itemID = itemID,
         itemSubTypeID = itemSubTypeID,
-        itemEquipLoc = itemEquipLoc
+        itemEquipLoc = itemEquipLoc,
+        sellPrice = sellPrice,
```
Comment (SabreValkyrn) [inline ItemUpgrades.lua:780]: Add `sellPrice` (used for vendor recommendation) and `itemMinLevel` (currently unused)

File: RXPGuides.lua [hunk 2]
```diff
@@ -165,32 +165,35 @@ local questFrame = CreateFrame("Frame");
...
-            if k == title or addon.GetQuestName(k) == title then
-                element = v
-            end
-        end
-        if element and element.step.active then
-            addon:SendEvent("RXP_QUEST_ACCEPT",element.questId)
-            return true
-        end
+function addon.QuestAutoAccept(titleOrId)
+    if not titleOrId then return end
```
Comment (SabreValkyrn) [inline RXPGuides.lua:169]: This appears to work on Classic, Wrath, and Retail.

Removed looping over entire `questAccept` as questID and questTitle are both keys, presumably weren't always.

Also removed nesting and added short circuits.

File: RXPGuides.lua [hunk 3]
```diff
@@ -533,9 +808,10 @@ function addon:QuestAutomation(event, arg1, arg2, arg3)
         local nActive = GetNumActiveQuests()
         local nAvailable = GetNumAvailableQuests()
 
+        local title, isComplete
         for i = 1, nActive do
-            local title, isComplete = GetActiveTitle(i)
-            if addon.QuestAutoTurnIn(title) and isComplete then
+            title, isComplete = GetActiveTitle(i)
+            if addon.GetStepQuestReward(title) and isComplete then
```
Comment (SabreValkyrn) [inline RXPGuides.lua:814]: Rename function to be a little more clear on what it does

File: RXPGuides.lua [hunk 4]
```diff
@@ -165,32 +165,35 @@ local questFrame = CreateFrame("Frame");
...
-        for k, v in pairs(addon.questTurnIn) do
-            if k == title or addon.GetQuestName(k) == title then
-                element = v
-            end
-        end
-        return addon.settings.profile.enableQuestRewardAutomation and element and
-                element.step.active and element.reward >= 0 and element.reward or 0
-    end
+function addon.GetStepQuestReward(titleOrId)
+    -- enableQuestRewardAutomation is setting for hard-coded .turnin step data
```
Comment (SabreValkyrn) [inline RXPGuides.lua:185]: Removed looping over entire `questTurnIn` as questID and questTitle are both keys, presumably weren't always.

Also removed nesting and added short circuits.

File: RXPGuides.lua [hunk 5]
```diff
@@ -478,6 +481,288 @@ local GossipSelectActiveQuest = C_GossipInfo.SelectActiveQuest or
...
+            if bestRatioFrame then
+                questRewardChoiceIcons["ratio"]:SetPoint("TOPRIGHT", bestRatioFrame , -1, 1)
+                questRewardChoiceIcons["ratio"]:SetParent(bestRatioFrame)
+                questRewardChoiceIcons["ratio"]:Show()
+            end
+        end
+    end
+
+    if addon.version < 40000 and addon.settings.profile.enableQuestChoiceGoldRecommendation then
+        local bestSellFrame = QuestInfo_GetRewardButton(QuestInfoFrame.rewardsFrame, bestSellOption)
```
Comment (SabreValkyrn) [inline RXPGuides.lua:677]: This is consistent between all 3 clients

File: RXPGuides.lua [hunk 6]
```diff
@@ -478,6 +481,288 @@ local GossipSelectActiveQuest = C_GossipInfo.SelectActiveQuest or
...
+        return
+    end
+
+    local title, _, _, _, _, _, _, questID = GetQuestLogTitle(questLogIndex)
+
+    -- options third return only used for handleQuestComplete
+    local bestSellOption, bestRatioOption, _ = evaluateQuestChoices(questID, numChoices, nil, GetQuestLogItemLink, GetQuestLogChoiceInfo)
+
+    if addon.settings.profile.enableQuestChoiceRecommendation then
+        local bestRatioFrame = _G['QuestLogItem' .. bestRatioOption] or
```
Comment (SabreValkyrn) [inline RXPGuides.lua:737]: This is inconsistent on Wrath/Retail vs Classic. `QuestInfo_GetRewardButton` returns `QuestInfoRewardsFrameQuestInfoItem` on all 3 clients but Classic in actuality uses `QuestLogItem`.

File: RXPGuides.lua [hunk 7]
```diff
@@ -478,6 +481,288 @@ local GossipSelectActiveQuest = C_GossipInfo.SelectActiveQuest or
...
+
+    local bestSellOption, bestRatioOption, options = evaluateQuestChoices(id, numChoices, GetQuestItemInfo, GetQuestItemLink)
+
+    if addon.version < 40000 and addon.settings.profile.enableQuestChoiceRecommendation then
+        if bestRatioOption > 0 then
+            local bestRatioFrame = QuestInfo_GetRewardButton(QuestInfoFrame.rewardsFrame, bestRatioOption)
+
+            if bestRatioFrame then
+                questRewardChoiceIcons["ratio"]:SetPoint("TOPRIGHT", bestRatioFrame , -1, 1)
+                questRewardChoiceIcons["ratio"]:SetParent(bestRatioFrame)
```
Comment (SabreValkyrn) [inline RXPGuides.lua:670]: Need to `:SetParent` to get icon compatible with scrollframe

File: RXPGuides.lua [hunk 8]
```diff
@@ -478,6 +481,288 @@ local GossipSelectActiveQuest = C_GossipInfo.SelectActiveQuest or
...
+
+    -- If auto rewards disabled, abort because not doing anything further
+    if not addon.settings.profile.enableQuestChoiceAutomation then return end
+
+    -- upgrade is more useful than selling
+    if bestRatioOption > 0 then
+        -- if isUsable, then automatically pick
+        -- If not usable but recommended then leave the window open for user decision
+        if options and options[bestRatioOption].isUsable then
+            print("GetQuestReward(bestRatioOption)", bestRatioOption)
```
Comment (SabreValkyrn) [inline RXPGuides.lua:703]: TODO restore `GetQuestReward` call, printing so I don't have to actually re-do quests everytime for testing.

File: RXPGuides.lua [hunk 9]
```diff
@@ -478,6 +481,288 @@ local GossipSelectActiveQuest = C_GossipInfo.SelectActiveQuest or
...
+            addon:SendEvent("RXP_QUEST_TURNIN", id, numChoices, bestRatioOption)
+        end
+    elseif bestSellOption > 0 then
+        print("GetQuestReward(bestSellOption)", bestSellOption)
```
Comment (SabreValkyrn) [inline RXPGuides.lua:707]: TODO restore `GetQuestReward` call, printing so I don't have to actually re-do quests everytime for testing.

File: RXPGuides.lua [hunk 10]
```diff
@@ -478,6 +481,288 @@ local GossipSelectActiveQuest = C_GossipInfo.SelectActiveQuest or
...
+        addon:SendEvent("RXP_QUEST_TURNIN", id, numChoices, bestSellOption)
+    end
+end
+
+local GetQuestLogSelection = _G.GetQuestLogSelection
+local GetNumQuestLogChoices = _G.GetNumQuestLogChoices -- C_QuestLog.GetLogIndexForQuestID
```
Comment (SabreValkyrn) [inline RXPGuides.lua:713]: Not doing this for Retail, so will probably remove C_QuestLog comment/calls

File: SettingsPanel.lua [hunk 11]
```diff
@@ -1826,6 +1828,14 @@ function addon.settings:CreateAceOptionsPanel()
                         end,
                         hidden = true -- TODO Zarant
                     },
+                    enableQuestChoiceGoldRecommendation = {
```
Comment (SabreValkyrn) [inline SettingsPanel.lua:1831]: Gold recommendation works without ItemUpgrades (e.g. Wrath), so move to general tips section

File: RXPGuides.lua (hunk 2 above)
Comment (Zarant) [reply RXPGuides.lua:169]: This is a leftover from a time where the retail API and classic API diverged, so the addon.GetQuestName() call was necessary to make it work on the old classic client

**Pull request merged**